File: hack/update-quarkus-platform.js
//import {App, Octokit} from "octokit";
//import xml2js from "xml2js";


const xml2js = require('xml2js');

const {Octokit} = require("octokit");
const https = require('https');
const fs = require('fs/promises');
const { spawn } = require('node:child_process');

const getLatestPlatform = () => {
    return new Promise((resolve, reject) => {
        const options = {
            hostname: 'code.quarkus.io',
            port: 443,
            path: '/api/platforms',
            method: 'GET',
        }
        https.request(options, res => {
            res.on('error', err => {
                reject(err)
            })
            let chunks = []
            res.on('data',chunk => {
                chunks.push(chunk)
            })
            res.on('end', () => {
                const data = JSON.parse(Buffer.concat(chunks).toString())
                resolve(data.platforms[0].streams[0].releases[0].quarkusCoreVersion)
            })
        }).end()
    });
}

const octokit = new Octokit({auth: process.env.GITHUB_TOKEN});
const [owner, repo] = process.env.GITHUB_REPOSITORY.split('/')

const prExists = async (pred) => {

    let page = 1
    const perPage = 10;

    while (true) {
        const resp = await octokit.rest.pulls.list({
            owner: owner,
            repo: repo,
            state: 'open',
            per_page: perPage,
            page: page
        })

        for (const e of resp.data) {
            if (pred(e)) {
                return true
            }
        }
        if (resp.data.length < perPage) {
            return false
        }
        page++
    }

}

const parseString = (text) => new Promise((resolve, reject) => {
    xml2js.parseString(text, {}, (err, res) => {
        if (err) {
            reject(err)
        }
        resolve(res)
    })
})

const platformFromPom = async (pomPath) => {
    const pomData = await fs.readFile(pomPath, { encoding: 'utf8' });
    const pom = await parseString(pomData)
    return pom.project.properties[0]['quarkus.platform.version'][0]
}

const cePomPath = "templates/quarkus/cloudevents/pom.xml"
const httpPomPath = "templates/quarkus/http/pom.xml"

const prepareBranch = async (branchName, prTitle) => {
    const script = `git config user.email "automation@knative.team" && \\
  git config user.name "Knative Automation" && \\
  git checkout -b "${branchName}" && \\
  make zz_filesystem_generated.go && \\
  git add "${cePomPath}" "${httpPomPath}" zz_filesystem_generated.go && \\
  git commit -m "${prTitle}" && \\
  git push --set-upstream origin "${branchName}"
`
    const subproc = spawn("sh", ["-c", script], {stdio: ['inherit', 'inherit', 'inherit']})

    return new Promise((resolve, reject) => {
        subproc.on('exit', code => {
            if (code === 0) {
                resolve()
                return
            }
            reject(new Error("cannot prepare branch: non-zero exit code"))
        })
    })
}

const updatePlatformInPom = async(pomPath, newPlatform) => {
    const pomData = await fs.readFile(pomPath, { encoding: 'utf8' });
    const newPomDate = pomData.replace(new RegExp('<quarkus.platform.version>[\\w.]+</quarkus.platform.version>', 'i'),
                    `<quarkus.platform.version>${newPlatform}</quarkus.platform.version>`)
    await fs.writeFile(pomPath, newPomDate)
}

const main = async () => {
    const latestPlatform = await getLatestPlatform()
    const prTitle = `chore: update Quarkus platform version to ${latestPlatform}`
    const branchName = `update-quarkus-platform-${latestPlatform}`
    const cePlatform = await platformFromPom(cePomPath)
    const httpPlatform = await platformFromPom(httpPomPath)

    if (cePlatform === latestPlatform && httpPlatform === latestPlatform) {
        console.log("Quarkus platform is up-to-date!")
        return
    }
    
    if (await prExists(({title}) => title === prTitle)) {
        console.log("The PR already exists!")
        return
    }

    await updatePlatformInPom(cePomPath, latestPlatform)
    await updatePlatformInPom(httpPomPath, latestPlatform)
    await prepareBranch(branchName, prTitle)
    await octokit.rest.pulls.create({
        owner: owner,
        repo: repo,
        title: prTitle,
        body: prTitle,
        base: 'main',
        head: `${owner}:${branchName}`,
    })
    console.log("The PR has been created!")

}

main().then(value => {
    console.log("OK!")
}).catch(reason => {
    console.log("ERROR: ", reason)
    process.exit(1)
})

File: hack/update-quarkus-platform.clj
(ns qbot
  (:require
    [clojure.java.io :as io]
    [org.httpkit.client :as client]
    [cheshire.core :as json]
    [clojure.data.xml :as xml]
    [clojure.string :as str]
    [clojure.java.shell :as shell]
    )
  (:import (java.io InputStream)))

(defn get-latest-platform
  "Get latest Quarkus platform version by accessing https://code.quarkus.io/api/platforms."
  []
  (-> (slurp "https://code.quarkus.io/api/platforms")
      (json/parse-string keyword)
      (:platforms)
      (first)
      (:streams)
      (first)
      (:releases)
      (first)
      (:quarkusCoreVersion)))

(defn get-current-platform
  "Gets current platform from pom.xml"
  [pom-path]
  (letfn [(search [{:keys [tag content]}]
            (if (.endsWith (str tag) "quarkus.platform.version")
              (first content)
              (->> content
                   (map #(search %))
                   (filter some?)
                   (first))))]
    (with-open [r (io/reader pom-path)]
      (search (xml/parse r)))))

(defn update-platform
  "Updates Quarkus platform in pom.xml. Naïve textual substitution is employed.
  XML parsing would be overkill here, and it could change formatting."
  [new-platform pom-path]
  (let [pom-updated (str/replace
                      (slurp pom-path)
                      #"<quarkus.platform.version>[\w.]+</quarkus.platform.version>"
                      (str "<quarkus.platform.version>" new-platform "</quarkus.platform.version>"))]
    (spit pom-path pom-updated)))

(def ce-pom "templates/quarkus/cloudevents/pom.xml")
(def http-pom "templates/quarkus/http/pom.xml")

(defn prepare-branch
  "Creates branch with updated platform and pushes it."
  [branch-name title]
  (let [script (format "git config user.email \"automation@knative.team\" && \\
  git config ser.name \"Knative Automation\" && \\
  git checkout -b '%s' && \\
  make zz_filesystem_generated.go && \\
  git add '%s' '%s' zz_filesystem_generated.go && \\
  git commit -m '%s' && \\
  git push --set-upstream origin '%s'"branch-name ce-pom http-pom title branch-name)
        {:keys [exit out err]} (shell/sh "sh" "-c" script)]
    (when (not= exit 0)
      (throw (ex-info "cannot prepare branch" {:err err :out out :command script})))))

(def ^:dynamic *gh-token* nil)

(defn gh-api-request
  "Helper function for making GitHub API requests.
  It JSON-fy request/response body, throws error and injects auth token."
  [method url opts]
  (let [opts (merge {:url url :method method :as :stream} opts)
        opts (update opts :headers (fn [headers]
                                     (merge {"Accept"        "application/vnd.github+json"
                                             "Authorization" (str "token " *gh-token*)}
                                            headers)))
        opts (update opts :body (fn [body]
                                  (if (and (some? body)
                                           (not (instance? InputStream body))
                                           (not (instance? String body)))
                                    (json/generate-string body)
                                    body)))
        {:keys [body status error]} @(client/request opts)
        body (with-open [r (io/reader body)]
               (json/parse-stream-strict r keyword))]

    (cond
      (some? error)
      (throw error)
      ;;
      (<= 200 status 299)
      body
      ;;
      :else
      (throw (ex-info "request error" body)))))

(defn list-prs
  [{:keys [repo state page per-page]}]
  (gh-api-request :get
                  (str "https://api.github.com/repos/" repo "/pulls")
                  {:query-params {:state    state
                                  :page     page
                                  :per_page per-page}}))

(defn pr-open?
  "Checks whether the PR with given title is already opened in repository."
  [{:keys [repo title]}]
  (let [all-prs (->> (range)
                     (drop 1)
                     (map #(list-prs {:repo repo, :state :open, :page %, :per-page 30}))
                     (take-while #(not (empty? %)))
                     (apply concat))]
    (->> all-prs
         (filter #(= (:title %) title))
         (first)
         (some?))))

(defn create-pr
  "Creates a PR."
  [{:keys [repo title body head base]}]
  (gh-api-request :post
                  (str "https://api.github.com/repos/" repo "/pulls")
                  {:body {:title title
                          :body  body
                          :base  base
                          :head  head}})
  )

(defn -main [& args]
  (binding [*gh-token* (System/getenv "GITHUB_TOKEN")]
    (let [latest-platform (get-latest-platform)
          branch-name (str "update-quarkus-platform-" latest-platform)
          title (str "chore: update Quarkus platform version to " latest-platform)
          gh-repo-owner (System/getenv "GITHUB_REPOSITORY_OWNER")
          gh-repo (System/getenv "GITHUB_REPOSITORY")]
      (cond
        (= (get-latest-platform)
           (get-current-platform ce-pom)
           (get-current-platform http-pom))
        (println "Quarkus platform is up-to-date!")
        ;;
        (pr-open? {:repo  gh-repo
                   :title title})
        (println "The PR is already opened!")
        ;;
        :else
        (do (update-platform latest-platform ce-pom)
            (update-platform latest-platform http-pom)
            (prepare-branch branch-name title)
            (create-pr {:repo  gh-repo
                        :title title
                        :body  title
                        :base  "main"
                        :head  (str gh-repo-owner ":" branch-name)})
            (println "The PR has been created!"))))))

(-main)
